lms/services/blackboard_api/service.py

```diff
@@ -53,12 +56,34 @@ def list_files(self, course_id):
         :raise ProxyAPIAccessTokenError: if we don't have a Blackboard API
             access token for the current user
         """
-        return self._http_service.get(
+        files = self._http_service.get(
             self._api_url(f"courses/uuid:{course_id}/resources"),
             oauth=True,
             schema=BlackboardListFilesSchema,
         ).validated_data
 
+        for file_ in files:
+            file_["id"] = f"blackboard://content-resource/{file_['id']}/"
```

Comment:
Something needs to take an ID that we received from the Blackboard Files API and wrap it in a `blackboard://...` URL for storage in our DB. And something needs to take that `blackboard://...` URL and unwrap just the ID from it, because that's what you need when getting the file's public URL from the Blackboard API.

I thought it would be nice if the service's `list_files()` did the wrapping and the service's `public_url()` did the unwrapping, because then the wrapping and unwrapping are next to each other in the same file, and knowledge about the `blackboard://...` URL scheme is encapsulated in this one file:

* When creating an assignment `BlackboardAPIClient.list_files()` returns a list of files with `blackboard://...` IDs to the frontend. These are opaque strings to the frontend. When the user selects a file the frontend posts the selected ID to the `configure_module_item()` view which saves it to the DB. The ID is an opaque string to the `configure_module_item()` view and the DB as well
* When launching an assignment we look up the `document_url` (which is the `blackboard://...` URL) in the DB and the `BlackboardAPIClient.public_url()` view unwraps it to just the ID for use with the Blackboard API for getting a public URL

The only place where the `blackboard://...` format leaks out is in `JSConfig` which has to look at it to know whether we're dealing with a Blackboard Files assignment:

https://github.com/hypothesis/lms/blob/1a0324d8e0e8213bb0f27c46a2561e20206a17af/lms/resources/_js_config.py#L65-L73

This could perhaps be moved into a `BlackboardAPIClient.is_blackboard_files_url(url)` method, and then knowledge of the `blackboard://...` format would be entirely encapsulated within `BlackboardAPIClient`.

Reply:
When slightly weird thing about this is that when our `list_files()` API returns a list of files to our frontend the `blackboard://...` URL is in the `id` fields of the files, whereas when it's stored in the `ModuleItemConfiguration` table in the DB the `blackboard://...` URL is in the `document_url` column. So sometimes it's an ID, sometimes it's a URL